File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/model/Language.java
package org.jetbrains.bsp.bazel.server.sync.model;

import io.vavr.collection.HashSet;
import io.vavr.collection.List;
import io.vavr.collection.Set;

public enum Language {
  SCALA("scala", HashSet.of(".scala")),
  JAVA("java", HashSet.of(".java")),
  KOTLIN("kotlin", HashSet.of(".kt"), HashSet.of(Language.JAVA.name)),
  // TODO https://youtrack.jetbrains.com/issue/BAZEL-25
  // CPP("cpp", HashSet.of(".C", ".cc", ".cpp", ".CPP", ".c++", ".cp", "cxx", ".h", ".hpp")),
  THRIFT("thrift", HashSet.of(".thrift"));

  private final String name;
  private final Set<String> extensions;
  private final Set<String> allNames;

  Language(String name, Set<String> extensions, Set<String> dependentNames) {
    this.name = name;
    this.extensions = extensions;
    this.allNames = dependentNames.add(name);
  }

  Language(String name, Set<String> extensions) {
    this(name, extensions, HashSet.of());
  }

  public static List<Language> all() {
    return List.of(values());
  }

  public String getName() {
    return name;
  }

  public Set<String> getExtensions() {
    return extensions;
  }

  public Set<String> getAllNames() {
    return allNames;
  }
}

File: server/src/main/kotlin/org/jetbrains/bsp/bazel/server/sync/model/Language.kt
package org.jetbrains.bsp.bazel.server.sync.model

enum class Language(
    val id: String,
    val extensions: Set<String>,
    val binaryTargets: Set<String> = hashSetOf(),
    dependentNames: Set<String> = hashSetOf(),
    val dependencyRegex: Regex? = null,
) {
    SCALA("scala", hashSetOf(".scala")),
    JAVA("java", hashSetOf(".java"), binaryTargets = setOf("java_binary")),
    KOTLIN("kotlin", hashSetOf(".kt"), setOf("kt_jvm_binary"), hashSetOf(JAVA.id)),
    CPP("cpp", hashSetOf(".C", ".cc", ".cpp", ".CPP", ".c++", ".cp", "cxx", ".h", ".hpp")),
    PYTHON("python", hashSetOf(".py")),
    THRIFT("thrift", hashSetOf(".thrift")),
    RUST("rust", hashSetOf(".rs")),
    ANDROID(
      "android",
      emptySet(),
      setOf("android_binary", "android_library"),
      hashSetOf(JAVA.id),
      // This should be removed once https://github.com/bazelbuild/rules_kotlin/issues/273 is fixed
      "@@rules_kotlin~.*//third_party:android_sdk".toRegex(),
    );

    val allNames: Set<String> = dependentNames + id

    companion object {
        private val ALL = values().toSet()
        fun all() = ALL
    }
}
